SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,276 @@
+		{
+			const ASpatialWorldSettings* WorldSettings = GetWorld() ? Cast<ASpatialWorldSettings>(GetWorld()->GetWorldSettings()) : nullptr;
+
+			if (WorldSettings == nullptr)
+			{
+				UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, WorldSettings should inherit from SpatialWorldSettings to get the load balancing strategy."));
+				UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+				DefaultLBStrategy->AddToRoot();
+				LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+				LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+				return;
+			}
+
+			// This will be uncommented after the next PR.
+			// For each Layer, add a LB Strategy for that layer.
+// 			TMap<FName, FLBLayerInfo> WorkerLBLayers = WorldSettings->WorkerLBLayers;
+// 
+// 			for (const TPair<FName, FLayerInfo>& Layer : Settings->WorkerLayers)
+// 			{
+// 				FName LayerKey = Layer.Key;
+// 
+// 				// Look through the WorldSettings to find the LBStrategy type for this layer.
+// 				if (!WorkerLBLayers.Contains(Layer.Key))
+// 				{
+// 					UE_LOG(LogLayeredLBStrategy, Error, TEXT("Layer %s does not have a defined LBStrategy in the WorldSettings. It will not be simulated."), *(LayerKey.ToString()));
+// 					continue;
+// 				}
+// 
+// 				UAbstractLBStrategy* LBStrategy = NewObject<UAbstractLBStrategy>(this, WorkerLBLayers[Layer.Key].LoadBalanceStrategy);
+// 				LayerNameToLBStrategy.Add(LayerKey, LBStrategy);
+// 				LBStrategy->Init();
+// 				LBStrategy->AddToRoot();
+// 
+// 				for (const TSoftClassPtr<AActor>& ClassPtr : Layer.Value.ActorClasses)
+// 				{
+// 					ClassPathToLayer.Add(ClassPtr, LayerKey);
+// 				}
+// 			}
+
+			// Finally, add the default layer.
+			if (WorldSettings->LoadBalanceStrategy == nullptr)
+			{
+				UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, there must be a LoadBalancing strategy set. Using a 1x1 grid."));
+				UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+				DefaultLBStrategy->AddToRoot();
+				LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+				LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+			}
+			else
+			{
+				UAbstractLBStrategy* DefaultLBStrategy = NewObject<UAbstractLBStrategy>(this, WorldSettings->LoadBalanceStrategy);
+				DefaultLBStrategy->AddToRoot();
+				LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+				LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+			}
+		}
+	}
+}
+
+void ULayeredLBStrategy::SetLocalVirtualWorkerId(VirtualWorkerId InLocalVirtualWorkerId)
+{
+	LocalVirtualWorkerId = InLocalVirtualWorkerId;
+	for (const auto& Elem : LayerNameToLBStrategy)
+	{
+		Elem.Value->SetLocalVirtualWorkerId(InLocalVirtualWorkerId);
+	}
+}
+
+bool ULayeredLBStrategy::ShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
```

Comment:
In some cases IsReady is used in a check. 
Is there a valid case where ShouldHaveAuthority can be called where the LBS is not ready?

Reply:
As we've gone through development, we've changed the way LB class initialization works a number of times, so this keeps wavering a bit. Sometimes startup actors can be created before LB classes are ready, sometimes it can't. I think we'll likely need to do a pass over all the LBStrategies before dev ready to have a consistent approach, but I don't think it's worthwhile to do now when things are still in flux.